### PR TITLE
Contract fixes

### PR DESCRIPTION
**Pull request by archseer**
[don't squash]

### CODE REVIEW

File: gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/setBilling.ts [hunk 1]
```diff
@@ -44,6 +46,15 @@ export default class SetBilling extends SolanaCommand {
...
 
     const info = await program.account.state.fetch(state)
+    const tokenVault = new PublicKey(info.config.tokenVault)
+    const [vaultAuthority] = await PublicKey.findProgramAddress(
+      [Buffer.from(utils.bytes.utf8.encode('vault')), state.toBuffer()],
+      program.programId,
+    )
+    const payees = info.oracles.xs
+      .slice(0, info.oracles.len)
+      .map((oracle) => ({ pubkey: oracle.payee, isWritable: true, isSigner: false }))
```
Comment (aalu1418) [inline gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/setBilling.ts:56]: thinking about if this works on a initialization edge case where oracles are not configured yet.
for example: the flow command follows set_billing => propose config
(i will test this out locally)

Comment (aalu1418) [reply gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/setBilling.ts:56]: verified this locally that `set_billing` still works when a config has not been set

File: contracts/programs/ocr2/src/context.rs [hunk 2]
```diff
@@ -136,12 +136,22 @@ pub struct RequestNewRound<'info> {
...
     pub authority: Signer<'info>,
     pub access_controller: AccountLoader<'info, AccessController>,
+
+    #[account(mut, address = state.load()?.config.token_vault)]
+    pub token_vault: Account<'info, TokenAccount>,
+    /// CHECK: This is a PDA
+    #[account(seeds = [b"vault", state.key().as_ref()], bump = state.load()?.vault_nonce)]
+    pub vault_authority: AccountInfo<'info>,
+
+    pub token_program: Program<'info, Token>,
```
[the remark this answers is not in the document]
Comment (aalu1418) [reply contracts/programs/ocr2/src/context.rs:154]: fixed! (e2e tests)

File: contracts/programs/store/src/lib.rs [hunk 3]
```diff
@@ -55,7 +55,21 @@ pub mod store {
...
+        let len = feed.to_account_info().data_len();
+        // discriminator + header size
+        let len = len
+            .checked_sub(8 + state::HEADER_SIZE)
+            .ok_or(ErrorCode::InsufficientSize)?;
+        require!(len % size_of::<Transmission>() == 0, InsufficientSize);
+        let space = len / size_of::<Transmission>();
+        // Live length must not exceed total capacity
+        require!(live_length <= space as u32, InvalidInput);
+
```
[the remark this answers is not in the document]
Comment (archseer) [reply contracts/programs/store/src/lib.rs:72]: Yeah good point 👍🏻

File: gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/setBilling.ts (hunk 1 above)
Comment (archseer) [reply gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/setBilling.ts:56]: Yep, in that case `oracles.len() == 0` so the contract also expects 0 `remaining_accounts`

**Review by krebernisak — DISMISSED** [review]

File: contracts/programs/ocr2/src/lib.rs [hunk 4]
```diff
@@ -600,6 +512,68 @@ pub mod ocr2 {
...
+                    authority: vault_authority.clone(),
+                },
+            );
+
+            Ok((amount_gjuels, cpi))
+        })
+        .collect::<Result<_>>()?;
+
+    // No account can be borrowed during CPI...
+    drop(state);
```
[the remark this answers is not in the document]
Comment (aalu1418) [reply contracts/programs/ocr2/src/lib.rs:561]: yup! for some context from a internal conversation:
> You mean "No account can be borrowed during CPI..."? Both cases work, I just explicitly drop(state) so it's evident it's dropped before CPI happens. In the other case the borrow checker does this.
> 
> We can share the code for both calls, it just requires some work because the ctx type isn't the same.

File: ops/solana/solana.go [hunk 5]
```diff
@@ -331,6 +331,24 @@ func (d Deployer) InitOCR(keys []opsChainlink.NodeKeys) error {
...
 
+	fmt.Println("Setting Billing...")
+	input := map[string]interface{}{
+		"observationPaymentGjuels":  1,
+		"transmissionPaymentGjuels": 1,
+	}
+	jsonInput, err := json.Marshal(input)
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (aalu1418) [reply ops/solana/solana.go:342]: we could but i think having a map[string] makes it easier in the future if additional parameters get added. it feels easier to read too rather than a single line string

File: contracts/programs/store/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (aalu1418) [reply contracts/programs/store/src/lib.rs:72]: added a test case!

File: ops/solana/solana.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (krebernisak) [reply ops/solana/solana.go:342]: Ah, it's debatable :) But makes sense to leave it for now, and not change just this one instance.

**Review by krebernisak — APPROVED** [review]
Approving for now, but would like to go back and investigate this suggestion:

- https://github.com/smartcontractkit/chainlink-solana/pull/243/files#r822396122

File: contracts/programs/ocr2/src/context.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (aalu1418) [reply contracts/programs/ocr2/src/context.rs:154]: captured the requests here: https://github.com/smartcontractkit/chainlink-solana/issues/247
it would be a larger body of work to standardize everything
